Issue/826 hide shipping virtual products

anitaa1990 opened the pull request:
Fixes #826 by hiding the **shipping** card for an order, if all the products in the order are virtual. This change takes place in both Order Detail and Order Fulfill screens.

### Note:
This [FluxC PR](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1287) must be merged before merging this one. I am keeping this PR in draft till the FluxC PR is merged.

### Tests
- [x] Run `OrderDetailPresenterTest` and `OrderFulfillmentPresenterTest`  unit tests
- [x] Run `OrderDetailCustomerInfoCardTest` and `OrderFulfillmentCustomerInfoCardTest`  UI tests


### Screenshot for an order with only virtual products:
<img width="300" src="https://user-images.githubusercontent.com/22608780/60489774-c7221980-9cc2-11e9-8d2f-96bca92b04e1.png">  <img width="300" src="https://user-images.githubusercontent.com/22608780/60489792-d2754500-9cc2-11e9-8c75-1b5f26765de9.png">

### Screenshot for an order with both virtual and non-virtual products:
<img width="300" src="https://user-images.githubusercontent.com/22608780/60489822-e6b94200-9cc2-11e9-854f-d16e2e6ecb56.png">  <img width="300" src="https://user-images.githubusercontent.com/22608780/60489833-eb7df600-9cc2-11e9-80da-8f87ae9ac8c1.png">

### Screenshot for an order with only non-virtual products:
<img width="300" src="https://user-images.githubusercontent.com/22608780/60489877-06e90100-9cc3-11e9-893b-01fddb1a1558.png">  <img width="300" src="https://user-images.githubusercontent.com/22608780/60489906-18caa400-9cc3-11e9-811d-88c58bd56c99.png">

Update release notes:

- [x] If there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.

anitaa1990 commented:
Accidentally, committed a different issue fix in the last commit 🤦‍♀ ! Reverted that and fixed any merge conflicts. PR is good for a review 👍

kyleaparker commented:
> @kyleaparker can you verify we should be hiding the entire card here? It would make more sense to me to just hide the shipping section of this card if the order contains only virtual products.

If billing, phone, and email are provided they should be displayed as part of the customer card

anitaa1990 commented:
Thank you for the clarification @kyleaparker and @AmandaRiu !! I fixed the logic to hide only the shipping card if the product is virtual and updated the screenshots in the PR description. PR ready for another round of review 👍

AmandaRiu commented:
@anitaa1990 I spent a little time playing around with the changes and have concerns over the following items. Some of these I'm not really sure how we should handle, so including @kyleaparker as well:

## Virtual only, only customer name provided
@kyleaparker how should these types of scenarios be handled? The customer name is provided at the top of Order Detail, there is no address, email or phone number to display. Should we hide the card?

<img src="https://user-images.githubusercontent.com/5810477/60555310-6104b900-9cf9-11e9-902e-c4293d163179.png" width="300"/>

## Virtual only, no customer info provided
@kyleaparker how should these types of scenarios be handled? Should we just hide the card?

<img src="https://user-images.githubusercontent.com/5810477/60555320-71b52f00-9cf9-11e9-8859-a0935edb55a2.png" width="300"/>

## Virtual + Physical, only customer name provided

@kyleaparker how should these types of scenarios be handled? The customer name is provided at the top of Order Detail, there is no address, email or phone number to display. Should we hide the card?

Current on the left, new version on the right
<img src="https://user-images.githubusercontent.com/5810477/60555346-93aeb180-9cf9-11e9-9783-2051811f1c06.png" width="300"/> <img src="https://user-images.githubusercontent.com/5810477/60555335-885b8600-9cf9-11e9-80d0-6269cbfac8ca.png" width="300"/>

## Virtual + Physical, no shipping address provided
Shipping address used to be copied from the billing address if it wasn't defined, but now, it's no longer doing this. We now have a section that doesn't display anything useful for shipping on the order detail view, and no shipping info on order fulfillment. I'm not sure how this should be handled though. Should we continue to copy the billing address? Should we just print out "No address provided" (I think that's what iOS was talking about doing). Also, for physical orders, should we still show the billing (or copied from billing) address on the order fulfillment screen? @kyleaparker 

Order detail on left, fulfillment on right
<img src="https://user-images.githubusercontent.com/5810477/60555587-d6bd5480-9cfa-11e9-8d7a-517733487ff1.png" width="300"/> <img src="https://user-images.githubusercontent.com/5810477/60555590-d8871800-9cfa-11e9-83d1-381b758a1fd8.png" width="300"/>

## Intermittent issues with views
@anitaa1990 I also ran into some strange UI states while testing. They aren't easily reproducible but maybe you can figure out what's happening.

No shipping address provided:
<img src="https://user-images.githubusercontent.com/5810477/60555401-d5d7f300-9cf9-11e9-971c-be32e854dc06.png" width="300"/>

Mix of physical and virtual products, card formatted as if had only virtual products
<img src="https://user-images.githubusercontent.com/5810477/60555563-b7bec280-9cfa-11e9-9f1a-6373d0a31de0.png" width="300"/>

anitaa1990 commented:
@AmandaRiu, thanks for the quick review. I wanted to add my 2 cents here 😊  

We recently made some changes to the shipping and billing details in this [PR](https://github.com/woocommerce/woocommerce-android/pull/1161) to handle empty shipping/billing details. So,
- If shipping and billing details are **not** available, we display a message `No address specified`.
- We only display shipping address if it is available. So if either name, address, city or country is available, this would be displayed. Otherwise, this section will be hidden.
- The same logic is followed for displaying billing address. If the billing address is not available, but the phone and email is available, then the billing address section will be hidden but the phone and email will be displayed. If no address, phone or email is available, then the entire billing section will be hidden. 

This logic is followed irrespective of the type of products in the order. I believe these changes align with how iOS currently handles it. 
### Screenshots for the above changes:
<img  src="https://user-images.githubusercontent.com/22608780/60067434-206cd480-9728-11e9-9897-617b94374192.png">  <img src="https://user-images.githubusercontent.com/22608780/60067447-337fa480-9728-11e9-80ba-b756032c0bb3.png">

That being said, I defer to @kyleaparker on how we should handle scenarios you mentioned, where only the shipping name or country info is available. But I believe this would need to be handled irrespective of the type of products in the order. Would it be ok to handle this scenario in a separate PR?

anitaa1990 commented:
Regarding the intermittent UI issues, I had some luck:

> Mix of physical and virtual products, card formatted as if had only virtual products

Ah, this is a bit more tricky. So the reason this happening is because the the virtual product in this order is available in the local db but the physical product is not available in the db. So the `isVirtualProduct` method is returning `true` even though the order contains both virtual and non virtual products. I've added a fix to this in 426cb7f. This basically checks if the size of the `lineItems` in an order matches the product count stored in the local cache before checking if the product is virtual or not.

Steps to reproduce this issue:
- Clear the local cache and login to the app again.
- Click on Order #3154. It has the same product as Order #3145. 
- Click back to return to the Order list page. Now click on Order #3145.
- Notice that bug you mentioned where the shipping details are not displayed even though there is a physical product in the list. 

>  I also ran into some strange UI states while testing. No shipping address provided:

Ah, this again is a tricky issue. So when the Order #3154 is first loaded, the product information is not available in the local db. But then we fetch the product information from the server api (to load the product images) and store this information in the local db. So when you click on the back button to return to the Order list screen and click on the same order again, it displays the correct information because the product info is now available in the local db. 

I added a **possible** fix for this in 7fb46aa. I've added logic to refresh the shipping details, once the product information becomes available. If the product info is still not available in the db for some reason, then the shipping address will be displayed for now 😞 

PR ready for another round 👍

kyleaparker commented:
>@kyleaparker how should these types of scenarios be handled? The customer name is provided at the top of Order Detail, there is no address, email or phone number to display. Should we hide the card?

For all of these examples, where does the "United States" come from? Is this just a default input? If so I would say we can hide the card.

>Shipping address used to be copied from the billing address if it wasn't defined, but now, it's no longer doing this.  

Unless it's a virtual product, I would have thought we would still copy over the address as the shipping address. Why would it no longer be doing this?

anitaa1990 commented:
> For all of these examples, where does the "United States" come from? Is this just a default input? If so I would say we can hide the card.

Hey @kyleaparker, in this scenario, only the country information of the address is available. As far as I know, we are not displaying a default input in the app. 

> Unless it's a virtual product, I would have thought we would still copy over the address as the shipping address. Why would it no longer be doing this?

Umm, I'm sorry for the confusion here but I think from this comment [here](https://github.com/woocommerce/woocommerce-android/pull/1161#issuecomment-502681339), we decided to hide the shipping address if it is not available and just display the billing info instead?. The logic present at the moment is:
- If any one of the shipping name, address or country information is available, the shipping card will be displayed. If none of those 3 information is available, we just hide the shipping card and display the billing information instead. 
- The same goes logic for billing information. If any one of the billing name, address or country information is available, the billing address section will be displayed. If no name, address or country is available, then this section is hidden. If billing phone/email information is available but the billing address is not, then those would be displayed while hiding the billing address section. 
- If no billing/ shipping information is available, then we display `No address specified` message.

AmandaRiu commented:
@kyleaparker It looks like we would get this information from the API if the user had entered an address for the customer and then later removed it, which is what I was doing during my testing. Looking at other tests the country would show up blank. 
```
"shipping": {
    "first_name": "",
    "last_name": "",
    "company": "",
    "address_1": "",
    "address_2": "",
    "city": "",
    "state": "",
    "postcode": "",
    "country": "US"
},
```
So if I'm understanding correctly if the country is available to be displayed even if there is no official address, show it (just in case it's supposed to be there), if not, hide it. If the only thing on the card is the customer's name (ie - no billing, shipping, phone, email), hide the card in the order detail view, but show a "no address specified" on the order fulfillment view? I'm wondering if we need some designs for these scenarios just because I see a lot of back and forth on them.

anitaa1990 commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/OrderDetailPresenter.kt:
```diff
@@ -204,6 +205,30 @@ class OrderDetailPresenter @Inject constructor(
+        // verify that the lineitem product is in the local cache and
+        // that the product count in the local cache matches the lineItem count.
+        val productModels = productStore.getProductsByRemoteIds(selectedSite.get(), remoteProductIds)
+        if (productModels.isNullOrEmpty() && productModels.count() != remoteProductIds.count()) {
```
👍 Done in d34f136. I also modified the same logic in `OrderFulfillmentPresenter`

anitaa1990 commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/OrderDetailCustomerInfoView.kt:
```diff
@@ -39,33 +38,9 @@ class OrderDetailCustomerInfoView @JvmOverloads constructor(ctx: Context, attrs:
-            customerInfo_viewMore.setOnCheckedChangeListener(null)
-        }
+        // show shipping section only for non virtual products or if shipping info available
+        val showShipping = !billingOnly && !isShippingInfoEmpty
```
That makes perfect sense 👌Modified this in faf8689

anitaa1990 commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/OrderDetailFragment.kt:
```diff
@@ -179,6 +182,12 @@ class OrderDetailFragment : androidx.fragment.app.Fragment(), OrderDetailContrac
 
+    override fun refreshCustomerInfoCard(order: WCOrderModel) {
+        // hide the shipping details if products in an order is virtual
+        val showShipping = !presenter.isVirtualProduct(order.getLineItemList())
```
Modified in faf8689

anitaa1990 commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/OrderDetailCustomerInfoView.kt:
```diff
@@ -116,6 +91,42 @@ class OrderDetailCustomerInfoView @JvmOverloads constructor(ctx: Context, attrs:
         }
     }
 
+    fun showShippingSection(order: WCOrderModel, show: Boolean) {
```
Modified in faf8689

anitaa1990 commented:
Thanks @AmandaRiu!! I've modified the nitpicks. PR ready for another round 👍

kyleaparker commented:
>@kyleaparker how should these types of scenarios be handled? The customer name is provided at the top of Order Detail, there is no address, email or phone number to display. Should we hide the card?

If no information is available at all I think we should hide it. I took a look how core handles this:
- Billing address is displayed even if only a country is selected
- Shipping address is display _only_ if an actual address is entered, if it's only the country then it's hidden

>@kyleaparker how should these types of scenarios be handled? The customer name is provided at the top of Order Detail, there is no address, email or phone number to display. Should we hide the card?

This would still display as it looks like billing address has a country recorded in that example 

>We now have a section that doesn't display anything useful for shipping on the order detail view, and no shipping info on order fulfillment. I'm not sure how this should be handled though. Should we continue to copy the billing address? Should we just print out "No address provided" (I think that's what iOS was talking about doing). Also, for physical orders, should we still show the billing (or copied from billing) address on the order fulfillment screen? @kyleaparker

It sounds like in that case the shipping address would have been manually deleted. I think if the customer specifies the billing address is the same as the shipping address that would automatically be copied on the order during checkout. So if a shipping address is manually deleted we don't need to show it. I don't think we need the billing address on the fulfilment screen as billing details aren't needed during fulfilment.

AmandaRiu commented:
@anitaa1990 I see that Kylea posted a bunch of answers after your last round of changes. Can you look them over and make any needed changes if necessary and let me know when this is ready for another review? Thanks!

anitaa1990 commented:
Thanks @kyleaparker for the clarifications here! Just to summarise the points we've discussed: 

#### Virtual products:
 - In Order Detail screen:
    - The shipping section would be hidden. 
    - The billing section would be displayed even if **only the country information is available**.
 - In Order Fulfill screen:
    - The shipping card will be hidden.

<img width="300" src="https://user-images.githubusercontent.com/22608780/60940477-403ff300-a2f9-11e9-9f67-74dd0f7d1ff5.png">  <img width="300" src="https://user-images.githubusercontent.com/22608780/60940480-433ae380-a2f9-11e9-8d2a-0a37bd25c7e7.png">


#### Virtual products + Physical products:
 - In both Order Detail/Order Fulfill screen:
    - Shipping address is displayed only if an actual address is entered, if it's only the country or name is provided, then it's hidden.
    - The billing section would be displayed even if **only the country information is available**.

##### When shipping address are available:
<img width="300" src="https://user-images.githubusercontent.com/22608780/60941061-
5cdd2a80-a2fb-11e9-9420-f710a4d8f9c5.png"> .  <img width="300" src="https://user-images.githubusercontent.com/22608780/60941078-6797bf80-a2fb-11e9-8f53-798dd84007aa.png">

##### When no shipping address are available:
<img width="300" src="https://user-images.githubusercontent.com/22608780/60941096-741c1800-a2fb-11e9-89cd-9ef1a45fccb5.png">  <img width="300" src="https://user-images.githubusercontent.com/22608780/60941120-8d24c900-a2fb-11e9-948e-a98649a95ca7.png">

@AmandaRiu I've fixed the merge conflicts and made some changes to hide the shipping card, if the shipping address is not available in Order Fulfill page 1cf3028. PR ready for another round 👍

kyleaparker commented:
Looks good!

This probably needs to be a separate issue, but an order with only virtual items shouldn't include shipment tracking.

anitaa1990 commented:
Thank you for the quick review @kyleaparker !!! I've created a separate issue [here](https://github.com/woocommerce/woocommerce-android/issues/1233) to address hiding shipment tracking card for virtual products.

@AmandaRiu PR ready for another round 👍
